clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +377,346 @@ impl FunctionType {
+                tracker,
+            )?;
+            inner_type_check_type(
+                db,
+                contract_context,
+                &atom_inner_types.1,
+                &expected_inner_types.1,
+                clarity_version,
+                depth + 1,
+                tracker,
+            )?;
+        }
+        (
+            TypeSignature::SequenceType(SequenceSubtype::ListType(atom_list_type)),
+            TypeSignature::SequenceType(SequenceSubtype::ListType(expected_list_type)),
+        ) => {
+            inner_type_check_type(
+                db,
+                contract_context,
+                atom_list_type.get_list_item_type(),
+                expected_list_type.get_list_item_type(),
+                clarity_version,
+                depth + 1,
+                tracker,
+            )?;
+        }
+        (
+            TypeSignature::TupleType(atom_tuple_type),
+            TypeSignature::TupleType(expected_tuple_type),
+        ) => {
+            if expected_tuple_type.get_type_map().len() != atom_tuple_type.get_type_map().len() {
+                return Err(
+                    CheckErrors::TypeError(expected_type.clone(), actual_type.clone()).into(),
+                );
+            }
+
+            for (name, expected_field_type) in expected_tuple_type.get_type_map() {
+                match atom_tuple_type.field_type(name) {
+                    Some(atom_field_type) => {
+                        inner_type_check_type(
+                            db,
+                            contract_context,
+                            atom_field_type,
+                            expected_field_type,
+                            clarity_version,
+                            depth + 1,
+                            tracker,
+                        )?;
+                    }
+                    None => {
+                        return Err(CheckErrors::TypeError(
+                            expected_type.clone(),
+                            actual_type.clone(),
+                        )
+                        .into())
                     }
                 }
             }
         }
-        Ok(returns.clone())
+        (
+            TypeSignature::CallableType(CallableSubtype::Trait(atom_trait_id)),
+            TypeSignature::CallableType(CallableSubtype::Trait(expected_trait_id)),
+        ) => {
+            if atom_trait_id != expected_trait_id {
+                let atom_trait = lookup_trait(
+                    db,
+                    contract_context,
+                    &atom_trait_id,
+                    clarity_version,
+                    tracker,
+                )?;
+                let expected_trait = lookup_trait(
+                    db,
+                    contract_context,
+                    expected_trait_id,
+                    clarity_version,
+                    tracker,
+                )?;
+                trait_check_trait_compliance(
```

Comment:
This should only be reachable when checking a Clarity2 contract, right?  Because, this function checks trait compliance differently than Clarity1.  What can we do to assert that this is the case (at compile-time if possible), so we don't accidentally use Clarity2 rules to evaluate Clarity1 code?

Reply:
That's right. In 7cd1293737682ff35125f2b1069cb7cbce408ff5, I refactored a bit and prefixed several function names with `clarity2_` as you suggested.